File: src/main/scala/com/teambr/bookshelf/client/gui/misc/Trackball.scala
package com.teambr.bookshelf.client.gui.misc

import java.lang.IllegalStateException

import com.google.common.base.Preconditions
import com.teambr.bookshelf.util.RenderUtils
import org.lwjgl.input.Mouse
import org.lwjgl.util.vector.{Matrix4f, Vector3f}

/**
  * This file was created for com.teambr.bookshelf.Bookshelf
  *
  * com.teambr.bookshelf.Bookshelf is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 1/31/2016
  */
class Trackball {

    var dragStart : Vector3f = _
    var lastTransform : Matrix4f = new Matrix4f()

    def calculateSpherePoint(x : Float, y : Float) : Vector3f = {
        val result = new Vector3f(x, y, 0)

        val sqrZ = 1 - Vector3f.dot(result, result)

        if(sqrZ > 0) result.z = Math.sqrt(sqrZ).toFloat
        else result.normalise()

        result
    }

    def getTransform(mouseX : Float, mouseY : Float) : Matrix4f = {
        if(dragStart == null) return lastTransform
        val current = calculateSpherePoint(mouseX, mouseY)

        val dot = Vector3f.dot(dragStart, current)
        if(Math.abs(dot - 1) < 0.0001) return lastTransform

        val axis = Vector3f.cross(dragStart, current, null)

        try {
            axis.normalise()
        } catch  { //Zero length vector
            case zeroLength : IllegalStateException => return lastTransform
            case e : Exception =>
        }

        val angle = 2.toFloat * Math.acos(dot).toFloat

        val rotation = new Matrix4f()
        rotation.rotate(angle, axis)
        Matrix4f.mul(rotation, lastTransform, null)
    }

    def applyTransform(mouseX : Float, mouseY : Float, isDragging : Boolean) : Unit = {
        RenderUtils.loadMatrix(if(isDragging) getTransform(mouseX, mouseY) else lastTransform)
    }

    def startDrag(mouseX : Float, mouseY : Float) : Unit = {
        dragStart = calculateSpherePoint(mouseX, mouseY)
    }

    def endDrag(mouseX : Float, mouseY : Float) : Unit = {
        lastTransform = getTransform(mouseX, mouseY)
    }
}

File: src/main/java/com/teambr/bookshelf/client/gui/misc/Trackball.java
package com.teambr.bookshelf.client.gui.misc;

import com.teambr.bookshelf.util.RenderUtils;
import org.lwjgl.util.vector.Matrix4f;
import org.lwjgl.util.vector.Vector3f;

/**
 * This file was created for Bookshelf
 * <p>
 * Bookshelf is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/13/2017
 */
public class Trackball {
    // Variables
    public Vector3f dragStart;
    public Matrix4f lastTransform = new Matrix4f();

    public Vector3f calculateSpherePoint(float x, float y) {
        Vector3f result = new Vector3f(x, y, 0);

        float sqrZ = 1 - Vector3f.dot(result, result);

        if(sqrZ > 0)
            result.z = (float) Math.sqrt(sqrZ);
        else
            result.normalise();

        return result;
    }

    public Matrix4f getTransform(float mouseX, float mouseY) {
        if(dragStart == null)
            return  lastTransform;

        Vector3f current = calculateSpherePoint(mouseX, mouseY);

        float dot = Vector3f.dot(dragStart, current);

        if(Math.abs(dot- 1) < 0.0001)
            return lastTransform;

        Vector3f axis = Vector3f.cross(dragStart, current, null);

        try {
            axis.normalise();
        } catch (IllegalStateException e){
            return lastTransform;
        } catch (Exception ignored) {}

        float angle = (float) (2 * Math.acos(dot));

        Matrix4f rotation = new Matrix4f();
        rotation.rotate(angle, axis);
        return Matrix4f.mul(rotation, lastTransform, null);
    }

    public void applyTransform(float mouseX, float mouseY, boolean isDragging) {
        RenderUtils.loadMatrix(isDragging ? getTransform(mouseX, mouseY) : lastTransform);
    }

    public void startDrag(float mouseX, float mouseY) {
        dragStart = calculateSpherePoint(mouseX, mouseY);
    }

    public void endDrag(float mouseX, float mouseY) {
        lastTransform = getTransform(mouseX, mouseY);
    }
}
